File: backend/src/middleware/error-handler.ts
import { type Request, type Response, type NextFunction } from "express";
import { AppError } from "../shared/utils/errors.js";
import { ZodError } from "zod";

export function errorHandler(err: Error, _req: Request, res: Response, _next: NextFunction) {
  if (err instanceof AppError) {
    res.status(err.statusCode).json({ error: err.message });
    return;
  }

  if (err instanceof ZodError) {
    const messages = err.errors.map((e) => e.message).join(", ");
    res.status(400).json({ error: messages });
    return;
  }

  console.error("Erro não tratado:", err);
  res.status(500).json({ error: "Erro interno do servidor" });
}

File: backend/src/middleware/error_handler.py
from fastapi import Request
from fastapi.responses import JSONResponse
from pydantic import ValidationError

from shared.utils.errors import AppError


async def app_error_handler(_request: Request, exc: AppError) -> JSONResponse:
    return JSONResponse(status_code=exc.status_code, content={"error": exc.message})


async def validation_error_handler(_request: Request, exc: ValidationError) -> JSONResponse:
    messages = ", ".join(e["msg"] for e in exc.errors())
    return JSONResponse(status_code=400, content={"error": messages})
